### PR TITLE
[libc][math][c23] Add {frexp,ilogb,llogb,logb,modf}f16 C23 math functions

### PR DESCRIPTION
**Pull request by overmighty**
Part of #93566.

### CODE REVIEW

**Comment by overmighty** [comment]
cc @lntue

File: libc/src/__support/FPUtil/ManipulationFunctions.h [hunk 1]
```diff
@@ -102,7 +103,7 @@ intlogb(U x) {
     return IntLogbConstants<T>::T_MAX;
   }
 
-  DyadicFloat<FPBits<U>::STORAGE_LEN> normal(bits.get_val());
+  DyadicFloat<cpp::max(FPBits<U>::STORAGE_LEN, 32)> normal(bits.get_val());
```
Comment (overmighty) [inline libc/src/__support/FPUtil/ManipulationFunctions.h:106]: `DyadicFloat<Bits>` uses `UInt<Bits>` which requires `Bits` to be either 32 or a multiple of the size of `WordType` which is never `uint16_t`.

File: libc/src/__support/FPUtil/dyadic_float.h [hunk 2]
```diff
@@ -183,6 +184,10 @@ template <size_t Bits> struct DyadicFloat {
     return r;
   }
 
+  LIBC_INLINE explicit constexpr operator float16() const {
+    return static_cast<float16>(static_cast<float>(*this));
+  }
```
Comment (overmighty) [inline libc/src/__support/FPUtil/dyadic_float.h:189]: The smoke test for `ldexpf16` showed that `operator T()` defined above did not always give correct results for `float16`. At least this part did not: https://github.com/llvm/llvm-project/blob/main/libc/src/__support/FPUtil/dyadic_float.h#L160-L161.

Maybe we could have an exhaustive test for `ldexpf16`.

File: libc/test/src/math/smoke/ILogbTest.h [hunk 3]
```diff
@@ -9,7 +9,7 @@
 #ifndef LLVM_LIBC_TEST_SRC_MATH_ILOGBTEST_H
 #define LLVM_LIBC_TEST_SRC_MATH_ILOGBTEST_H
 
-#include "src/__support/CPP/limits.h" // INT_MAX
```
Comment (overmighty) [inline libc/test/src/math/smoke/ILogbTest.h:12]: Was unused. I did similar cleanups in test templates that I edited in previous PRs, but I don't think I ever left a comment explaining it.

File: libc/test/src/math/smoke/ILogbTest.h [hunk 4]
```diff
@@ -94,10 +96,12 @@ class LlvmLibcILogbTest : public LIBC_NAMESPACE::testing::FEnvSafeTest {
   void test_normal_range(Func func) {
     constexpr StorageType MIN_NORMAL = FPBits::min_normal().uintval();
     constexpr StorageType MAX_NORMAL = FPBits::max_normal().uintval();
-    constexpr StorageType COUNT = 10'001;
-    constexpr StorageType STEP = (MAX_NORMAL - MIN_NORMAL) / COUNT;
+    constexpr int COUNT = 10'001;
+    constexpr StorageType STEP = LIBC_NAMESPACE::cpp::max(
+        static_cast<StorageType>((MAX_NORMAL - MIN_NORMAL) / COUNT),
+        StorageType(1));
```
Comment (overmighty) [inline libc/test/src/math/smoke/ILogbTest.h:102]: Here, `COUNT` was small enough and `MAX_NORMAL - MIN_NORMAL` large enough for `STEP` to be > 0 for `float16`/`uint16_t`, but I thought it wouldn't hurt to also do the `max(..., 1)` change here anyway.

File: libc/test/src/math/smoke/LdExpTest.h [hunk 5]
```diff
@@ -31,7 +31,7 @@ class LdExpTestTemplate : public LIBC_NAMESPACE::testing::FEnvSafeTest {
   const T nan = FPBits::quiet_nan().get_val();
 
   // A normalized mantissa to be used with tests.
-  static constexpr StorageType MANTISSA = NormalFloat::ONE + 0x1234;
+  static constexpr StorageType MANTISSA = NormalFloat::ONE + 0x123;
```
Comment (overmighty) [inline libc/test/src/math/smoke/LdExpTest.h:34]: `0x1234` is 13 bits, too large for `float16`'s 10-bit fraction.

File: libc/test/src/math/smoke/LdExpTest.h [hunk 6]
```diff
@@ -60,7 +60,7 @@ class LdExpTestTemplate : public LIBC_NAMESPACE::testing::FEnvSafeTest {
 
   void testOverflow(LdExpFunc func) {
     NormalFloat x(Sign::POS, FPBits::MAX_BIASED_EXPONENT - 10,
-                  NormalFloat::ONE + 0xF00BA);
+                  NormalFloat::ONE + 0xF00);
```
Comment (overmighty) [inline libc/test/src/math/smoke/LdExpTest.h:63]: `0xF00BA` is 20 bits, too large for `float16`'s 10-bit fraction.

File: libc/test/src/math/smoke/LdExpTest.h [hunk 7]
```diff
@@ -95,10 +95,10 @@ class LdExpTestTemplate : public LIBC_NAMESPACE::testing::FEnvSafeTest {
...
   void testNormalOperation(LdExpFunc func) {
     T val_array[] = {// Normal numbers
-                     NormalFloat(Sign::POS, 100, MANTISSA),
-                     NormalFloat(Sign::POS, -100, MANTISSA),
-                     NormalFloat(Sign::NEG, 100, MANTISSA),
-                     NormalFloat(Sign::NEG, -100, MANTISSA),
+                     NormalFloat(Sign::POS, 10, MANTISSA),
+                     NormalFloat(Sign::POS, -10, MANTISSA),
+                     NormalFloat(Sign::NEG, 10, MANTISSA),
+                     NormalFloat(Sign::NEG, -10, MANTISSA),
```
Comment (overmighty) [inline libc/test/src/math/smoke/LdExpTest.h:101]: Exponents -100 and 100 are too large for `float16`.

File: libc/src/__support/FPUtil/ManipulationFunctions.h (hunk 1 above)
Comment (lntue) [reply libc/src/__support/FPUtil/ManipulationFunctions.h:106]: Maybe update `DyadicFloat` to accept `Bits` to also accept 16?

**Comment by overmighty** [comment]
Rebased and resolved merge conflitcts.

**Review by lntue — APPROVED** [review]

**Comment by overmighty** [comment]
Wait, I haven't pushed a change to address this yet:

> Maybe update `DyadicFloat` to accept `Bits` to also accept 16?

**Comment by lntue** [comment]
> Wait, I haven't pushed a change to address this yet:
> 
> > Maybe update `DyadicFloat` to accept `Bits` to also accept 16?

A separate PR is better.

**Pull request merged**